### PR TITLE
 First Submit react week3 homework

### PR DESCRIPTION
**Pull request by shanhaider23**

### CODE REVIEW

File: react/week3/todoapp-w3/src/components/Todo.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    value: "",
+  });
+
+  const submitUpdate = (value) => {
+    updateTodo(edit.id, value);
+    setEdit({
+      id: null,
+      value: "",
+    });
+  };
```
Comment (calm-prog) [inline react/week3/todoapp-w3/src/components/Todo.js:18]: I think it's really smart that you directly edit the value here :+1: 
Others got into trouble trying to modify the todos array.

File: react/week3/todoapp-w3/src/components/TodoList.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import Todo from "./Todo";
+import FetchAPI from "./Fetch";
+import TimeUsed from "./TimeUsed";
+import Deadline from "./DatePicker";
+import PropTypes from "prop-types";
+function TodoList() {
+  const [todos, setTodos] = useState([]);
+
+  const addTodo = (todo) => {
+    if (!todo.text || /^\s*$/.test(todo.text)) {
```
Comment (calm-prog) [inline react/week3/todoapp-w3/src/components/TodoList.js:12]: Proper input validation, that's cool!

File: react/week3/todoapp-w3/src/components/TodoList.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  const updateTodo = (todoId, newValue) => {
+    if (!newValue.text || /^\s*$/.test(newValue.text)) {
+      return;
+    }
+
+    setTodos((prev) =>
+      prev.map((item) => (item.id === todoId ? newValue : item))
+    );
+  };
```
Comment (calm-prog) [inline react/week3/todoapp-w3/src/components/TodoList.js:30]: To me this logic is so neat! Awesome

**Review by calm-prog — COMMENTED** [review]
Wow, that's next level! Not only you understand the declarative logic of React, but you write your code so elegantly and organize the logic so smart in my opinion. It was pure joy reading it. 

Just one thing to note is that probably from a user point of view it would be more intuitive if the todo task and the todo deadline could be submitted together and they would also appear in a single component.

Anyway, of course there is flexibility what behavior you want to define for your app. The code itself is super good, I have only positive comments.

File: react/week3/todoapp-w3/src/components/TodoList.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    const removedArr = [...todos].filter((todo) => todo.id !== id);
+
+    setTodos(removedArr);
+  };
+
+  const completeTodo = (id) => {
+    let updatedTodos = todos.map((todo) => {
+      if (todo.id === id) {
+        todo.isComplete = !todo.isComplete;
+      }
```
Comment (calm-prog) [inline react/week3/todoapp-w3/src/components/TodoList.js]: The problem is here that even though you are creating a new array with the map() function, you are still modifying the original object with the assignment operator. 

An immutable way of doing this is using the spread operator here:
```
if(todo.id === id) {
  return {...todo, isComplete: !isComplete}
} else {
  return todo
}
```

So with the spread operator first I copy the todo with the 3 dots, and after the comma I specify the key whose value I want to update.

**Review by calm-prog — COMMENTED** [review]
Update: I was looking at another homework when I realized that there is an immutability issue inside the completeTodo function. I commented in the code below.

File: react/week3/todoapp-w3/src/components/TodoList.js (hunk 4 above)
Comment (shanhaider23) [reply react/week3/todoapp-w3/src/components/TodoList.js]: Thanks for the comments. I have changed the required code

Comment (calm-prog) [reply react/week3/todoapp-w3/src/components/TodoList.js]: Great, it looks good. :+1: You can merge it to master.

**Pull request merged**